File: QRScanner/QRScanner/DTCameraPreviewController.m
//
//  ViewController.m
//  Camera
//
//  Created by Oliver Drobnik on 05.11.13.
//  Copyright (c) 2013 Cocoanetics. All rights reserved.
//

#import "DTCameraPreviewController.h"

#import "DTAVFoundationFunctions.h"
#import "DTVideoPreviewView.h"


// private interface to tag with metadata delegate protocol
@interface DTCameraPreviewController () <AVCaptureMetadataOutputObjectsDelegate>
@end

@implementation DTCameraPreviewController
{
	AVCaptureDevice *_camera;
	AVCaptureDeviceInput *_videoInput;
	AVCaptureStillImageOutput *_imageOutput;
	AVCaptureSession *_captureSession;
	DTVideoPreviewView *_videoPreview;
	
	AVCaptureMetadataOutput *_metaDataOutput;
	dispatch_queue_t _metaDataQueue;
	
	NSMutableSet *_visibleCodes;
	NSMutableDictionary *_visibleCodeShapes;
}


- (void)dealloc
{
	[[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark - Internal Methods

- (void)_informUserAboutCamNotAuthorized
{
	UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Cam Access" message:@"Access to the camera hardware has been disabled. This disables all related functionality in this app. Please enable it via device settings." delegate:nil cancelButtonTitle:@"Ok" otherButtonTitles:nil];
	[alert show];
}

- (void)_informUserAboutNoCam
{
	UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Cam Access" message:@"The current device does not have any cameras installed, are you running this in iOS Simulator?" delegate:nil cancelButtonTitle:@"Yes" otherButtonTitles:nil];
	[alert show];
}


- (void)_setupMetadataOutput
{
	// create a new metadata output
	_metaDataOutput = [[AVCaptureMetadataOutput alloc] init];
	
	// create a GCD queue on which recognized codes are to be delivered
	//_metaDataQueue = dispatch_queue_create("com.cocoanetics.metadata", NULL);
	_metaDataQueue = dispatch_get_main_queue();

	// set self as delegate, using the background queue
	[_metaDataOutput setMetadataObjectsDelegate:self queue:_metaDataQueue];
	
	// connect meta data output only if possible
	if (![_captureSession canAddOutput:_metaDataOutput])
	{
		NSLog(@"Unable to add meta data output to capture session");
		return;
	}
	
	// connect it
	[_captureSession addOutput:_metaDataOutput];
	 
	 // specify to scan for supported 2D barcode types
	NSArray *barcodes2D = @[AVMetadataObjectTypePDF417Code, AVMetadataObjectTypeQRCode, AVMetadataObjectTypeAztecCode];
	NSArray *availableTypes = [_metaDataOutput availableMetadataObjectTypes];
	
	if (![availableTypes count])
	{
		NSLog(@"Unable to get any available metadata types, did you forget the addOutput: on the capture session?");
		return;
	}
	
	// be extra defensive: only adds supported types, log unsupported
	NSMutableArray *tmpArray = [NSMutableArray array];
	
	for (NSString *oneCodeType in barcodes2D)
	{
		if ([availableTypes containsObject:oneCodeType])
		{
			[tmpArray addObject:oneCodeType];
		}
		else
		{
			NSLog(@"Weird: Code type '%@' is not reported as supported on this device", oneCodeType);
		}
	}
	
	_metaDataOutput.metadataObjectTypes = tmpArray;
	
	if ([tmpArray count])
	{
		 _metaDataOutput.metadataObjectTypes = tmpArray;
	}
}

- (void)_setupCamera
{
	// get the camera
	_camera = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
	
	if (!_camera)
	{
		[self.snapButton setTitle:@"No Camera Found" forState:UIControlStateNormal];
		self.snapButton.enabled = NO;
		
		[self _informUserAboutNoCam];
		
		return;
	}
	
	// connect camera to input
	NSError *error;
	_videoInput = [[AVCaptureDeviceInput alloc] initWithDevice:_camera error:&error];
	
	if (!_videoInput)
	{
		NSLog(@"Error connecting video input: %@", [error localizedDescription]);
		return;
	}

	// Create session (use default AVCaptureSessionPresetHigh)
	_captureSession = [[AVCaptureSession alloc] init];

	if (![_captureSession canAddInput:_videoInput])
	{
		NSLog(@"Unable to add video input to capture session");
		return;
	}
	
	[_captureSession addInput:_videoInput];

	// configure cam here because active format depends on capture session
	[self _configureCurrentCamera];
	
	// add still image output
	_imageOutput = [AVCaptureStillImageOutput new];
	
	if (![_captureSession canAddOutput:_imageOutput])
	{
		NSLog(@"Unable to add still image output to capture session");
		return;
	}
	
	[_captureSession addOutput:_imageOutput];
	
	// set the session to be previewed
	_videoPreview.previewLayer.session = _captureSession;
	
	// setup the barcode scanner output
	[self _setupMetadataOutput];
}

- (void)_setupCameraAfterCheckingAuthorization
{
	if (![[AVCaptureDevice class] respondsToSelector:@selector(authorizationStatusForMediaType:)])
	{
		// running on iOS 6, assume authorization
		[self _setupCamera];
		
		return;
	}
	
	AVAuthorizationStatus authStatus = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo];
	
	switch (authStatus)
	{
		case AVAuthorizationStatusAuthorized:
		{
			[self _setupCamera];
			
			break;
		}
			
		case AVAuthorizationStatusNotDetermined:
		{
			[AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo completionHandler:^(BOOL granted) {
				
				// background thread, we want to setup on main thread
				dispatch_async(dispatch_get_main_queue(), ^{
					
					if (granted)
					{
						[self _setupCamera];
						
						// need to do more setup because View Controller is already present
						[self _setupCamSwitchButton];
						[self _setupTorchToggleButton];
						
						// need to start capture session
						[_captureSession startRunning];
					}
					else
					{
						[self _informUserAboutCamNotAuthorized];
					}
				});
			}];
			
			break;
		}

		case AVAuthorizationStatusRestricted:
		case AVAuthorizationStatusDenied:
		{
			[self _informUserAboutCamNotAuthorized];
			
			break;
		}
	}
}

// applies settings to the currently active camera
- (void)_configureCurrentCamera
{
	// if cam supports AV lock then we want to be able to get out of this
	if ([_camera isFocusModeSupported:AVCaptureFocusModeLocked])
	{
		if ([_camera lockForConfiguration:nil])
		{
			_camera.subjectAreaChangeMonitoringEnabled = YES;
			
			if ([_camera isSmoothAutoFocusSupported])
			{
				_camera.smoothAutoFocusEnabled = YES;
			}
			
			// get more pixels to image outputs
			_camera.videoZoomFactor = MIN(_camera.activeFormat.videoZoomFactorUpscaleThreshold, 1.25);
			
			[_camera unlockForConfiguration];
		}
	}
}

// checks if there is an alternative camera to the current one
- (AVCaptureDevice *)_alternativeCamToCurrent
{
	if (!_camera)
	{
		// no current camera
		return nil;
	}
	
	NSArray *allCams = [AVCaptureDevice devicesWithMediaType:AVMediaTypeVideo];
	
	for (AVCaptureDevice *oneCam in allCams)
	{
		if (oneCam != _camera)
		{
			// found an alternative!
			return oneCam;
		}
	}
	
	// no alternative cameras present
	return nil;
}

- (AVCaptureConnection *)_captureConnection
{
	for (AVCaptureConnection *connection in _imageOutput.connections)
	{
		for (AVCaptureInputPort *port in [connection inputPorts])
		{
			if ([[port mediaType] isEqual:AVMediaTypeVideo] )
			{
				return connection;
				break;
			}
		}
	}
	
	// no connection found
	return nil;
}

// update all capture connections for the current interface orientation
- (void)_updateConnectionsForInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
	AVCaptureVideoOrientation captureOrientation = DTAVCaptureVideoOrientationForUIInterfaceOrientation(interfaceOrientation);
	
	for (AVCaptureConnection *connection in _imageOutput.connections)
	{
		if ([connection isVideoOrientationSupported])
		{
			connection.videoOrientation = captureOrientation;
		}
	}
	
	if ([_videoPreview.previewLayer.connection isVideoOrientationSupported])
	{
		_videoPreview.previewLayer.connection.videoOrientation = captureOrientation;
	}
}

// configures cam switch button
- (void)_setupCamSwitchButton
{
	AVCaptureDevice *alternativeCam = [self _alternativeCamToCurrent];
	
	if (alternativeCam)
	{
		self.switchCamButton.hidden = NO;

		NSString *title;
		
		switch (alternativeCam.position)
		{
			case AVCaptureDevicePositionBack:
			{
				title = @"Back";
				break;
			}
				
			case AVCaptureDevicePositionFront:
			{
				title = @"Front";
				break;
			}

			case AVCaptureDevicePositionUnspecified:
			{
				title = @"Other";
				break;
			}
		}
		
		[self.switchCamButton setTitle:title forState:UIControlStateNormal];
	}
	else
	{
		self.switchCamButton.hidden = YES;
	}
}

// configure torch button for current cam
- (void)_setupTorchToggleButton
{
	if ([_camera hasTorch])
	{
		self.toggleTorchButton.hidden = NO;
	}
	else
	{
		self.toggleTorchButton.hidden = YES;
	}
}

#pragma mark - View Appearance

- (void)viewDidLoad
{
    [super viewDidLoad];

	NSAssert([self.view isKindOfClass:[DTVideoPreviewView class]], @"Wrong root view class %@ in %@", NSStringFromClass([self.view class]), NSStringFromClass([self class]));
	
	_videoPreview = (DTVideoPreviewView *)self.view;
	
	// default is resize aspect, we need aspect fill to avoid side bars on iPad
	[_videoPreview.previewLayer setVideoGravity:AVLayerVideoGravityResizeAspectFill];
	
	[self _setupCameraAfterCheckingAuthorization];
	
	UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleTap:)];
	[self.view addGestureRecognizer:tap];
	
	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(subjectChanged:) name:AVCaptureDeviceSubjectAreaDidChangeNotification object:nil];
}

- (void)viewWillAppear:(BOOL)animated
{
	[super viewWillAppear:animated];
	
	// need to update capture and preview connections
	[self _updateConnectionsForInterfaceOrientation:self.interfaceOrientation];
	
	// start session so that we don't see a black rectangle, but video
	[_captureSession startRunning];
	
	[self _setupCamSwitchButton];
	[self _setupTorchToggleButton];
	
	_visibleCodes = [NSMutableSet new];
	_visibleCodeShapes = [NSMutableDictionary new];
}

- (void)viewDidDisappear:(BOOL)animated
{
	[super viewDidDisappear:animated];
	
	[_captureSession stopRunning];
}

- (BOOL)shouldAutorotate
{
	if ([_videoPreview.previewLayer.connection isVideoOrientationSupported])
	{
		return YES;
	}
	
	// prevent UI autorotation to avoid confusing the preview layer
	return NO;
}

#pragma mark - Interface Rotation

// for demonstration all orientations are supported
- (NSUInteger)supportedInterfaceOrientations
{
	return UIInterfaceOrientationMaskAll;
}

- (void)willRotateToInterfaceOrientation:(UIInterfaceOrientation)toInterfaceOrientation duration:(NSTimeInterval)duration
{
	[super willRotateToInterfaceOrientation:toInterfaceOrientation duration:duration];
	
	// need to update capture and preview connections
	[self _updateConnectionsForInterfaceOrientation:toInterfaceOrientation];
}

#pragma mark - AVCaptureMetadataOutputObjectsDelegate

- (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputMetadataObjects:(NSArray *)metadataObjects fromConnection:(AVCaptureConnection *)connection
{
	// set to take on codes that this pass of the method is reporting
	NSMutableSet *reportedCodes = [NSMutableSet set];
	
	// dictionary to count the number of occurences of a type+stringValue
	NSMutableDictionary *reportedCodesCount = [NSMutableDictionary dictionary];
	
	for (AVMetadataMachineReadableCodeObject *object in metadataObjects)
	{
		if ([object isKindOfClass:[AVMetadataMachineReadableCodeObject class]] && object.stringValue)
		{
			NSString *code = [NSString stringWithFormat:@"%@:%@", object.type, object.stringValue];
			
			// get the number of times this code was reported before in this loop
			NSUInteger occurencesOfCode = [reportedCodesCount[code] unsignedIntegerValue] + 1;
			reportedCodesCount[code] = @(occurencesOfCode);
			NSString *numberedCode = [code stringByAppendingFormat:@"-%lu", (unsigned long)occurencesOfCode];

			// if it was not previously visible it is new
			if (![_visibleCodes containsObject:numberedCode])
			{
				NSLog(@"code appeared: %@", numberedCode);
				
				if ([_delegate respondsToSelector:@selector(previewController:didScanCode:ofType:)])
				{
					[_delegate previewController:self didScanCode:object.stringValue ofType:object.type];
				}
			}

			[reportedCodes addObject:numberedCode];

			// create a suitable CGPath for the barcode area
			CGPathRef path = DTAVMetadataMachineReadableCodeObjectCreatePathForCorners(_videoPreview.previewLayer, object);
			
			// get previous shape for this code
			CAShapeLayer *shapeLayer = _visibleCodeShapes[numberedCode];
			
			// if none found then this is a new shape
			if (!shapeLayer)
			{
				shapeLayer = [CAShapeLayer layer];
				
				// basic configuration, stays the same regardless of path
				shapeLayer.strokeColor = [UIColor greenColor].CGColor;
				shapeLayer.fillColor = [UIColor colorWithRed:0 green:1 blue:0 alpha:0.25].CGColor;
				shapeLayer.lineWidth = 2;

				[_videoPreview.layer addSublayer:shapeLayer];
				
				// add it to shape dictionary
				_visibleCodeShapes[numberedCode] = shapeLayer;
			}
			
			// configure shape, relative to video preview
			shapeLayer.frame = _videoPreview.bounds;
			shapeLayer.path = path;
			
			// need to release the path now
			CGPathRelease(path);
		}
	}
	
	// check which codes which we saw in previous cycle are no longer present
	for (NSString *oneCode in _visibleCodes)
	{
		if (![reportedCodes containsObject:oneCode])
		{
			NSLog(@"code disappeared: %@", oneCode);
			
			CAShapeLayer *shape = _visibleCodeShapes[oneCode];
			
			[shape removeFromSuperlayer];
			[_visibleCodeShapes removeObjectForKey:oneCode];
		}
	}
	
	_visibleCodes = reportedCodes;
}


#pragma mark - Actions

- (IBAction)snap:(UIButton *)sender
{
	if (!_camera)
	{
		return;
	}
	
	// find correct connection
	AVCaptureConnection *videoConnection = [self _captureConnection];
	
	if (!videoConnection)
	{
		NSLog(@"Error: No Video connection found on still image output");
		return;
	}
	
	[_imageOutput captureStillImageAsynchronouslyFromConnection:videoConnection completionHandler:^(CMSampleBufferRef imageSampleBuffer, NSError *error) {
		
		if (error)
		{
			NSLog(@"Error capturing still image: %@", [error localizedDescription]);
			return;
		}
		
		NSData *imageData = [AVCaptureStillImageOutput jpegStillImageNSDataRepresentation:imageSampleBuffer];
		UIImage *image = [UIImage imageWithData:imageData];
		
		UIImageWriteToSavedPhotosAlbum(image, nil, nil, nil);
	 }];
}

- (IBAction)switchCam:(UIButton *)sender
{
	[_captureSession beginConfiguration];
	
	_camera = [self _alternativeCamToCurrent];
	
	// remove all old inputs
	for (AVCaptureDeviceInput *input in _captureSession.inputs)
	{
		[_captureSession removeInput:input];
	}
	
	// add new input
	_videoInput = [AVCaptureDeviceInput deviceInputWithDevice:_camera error:nil];
	[_captureSession addInput:_videoInput];
	
	// there are new connections, tell them about current UI orientation
	[self _updateConnectionsForInterfaceOrientation:self.interfaceOrientation];
	
	[_captureSession commitConfiguration];
	
	// configure camera after session changes
	[self _configureCurrentCamera];
	
	// update the buttons
	[self _setupCamSwitchButton];
	[self _setupTorchToggleButton];
}

- (IBAction)toggleTorch:(UIButton *)sender
{
	if ([_camera hasTorch])
	{
		BOOL torchActive = [_camera isTorchActive];
		
		// need to lock, without this there is an exception
		if ([_camera lockForConfiguration:nil])
		{
			if (torchActive)
			{
				if ([_camera isTorchModeSupported:AVCaptureTorchModeOff])
				{
					[_camera setTorchMode:AVCaptureTorchModeOff];
				}
			}
			else
			{
				if ([_camera isTorchModeSupported:AVCaptureTorchModeOn])
				{
					[_camera setTorchMode:AVCaptureTorchModeOn];
				}
			}
			
			[_camera unlockForConfiguration];
		}
	}
}

- (void)handleTap:(UITapGestureRecognizer *)gesture
{
	if (gesture.state == UIGestureRecognizerStateRecognized)
	{
		// require both focus point and autofocus
		if (![_camera isFocusPointOfInterestSupported] || ![_camera isFocusModeSupported:AVCaptureFocusModeAutoFocus])
		{
			NSLog(@"Focus Point Not Supported by current camera");
			
			return;
		}
		
		CGPoint locationInPreview = [gesture locationInView:_videoPreview];
		CGPoint locationInCapture = [_videoPreview.previewLayer captureDevicePointOfInterestForPoint:locationInPreview];
		
		if ([_camera lockForConfiguration:nil])
		{
			// this alone does not trigger focussing
			[_camera setFocusPointOfInterest:locationInCapture];
			
			// this focusses once and then changes to locked
			[_camera setFocusMode:AVCaptureFocusModeAutoFocus];
			
			NSLog(@"Focus Mode: Locked to Focus Point");

			[_camera unlockForConfiguration];
		}
	}
}

#pragma mark - Notifications

- (void)subjectChanged:(NSNotification *)notification
{
	// switch back to continuous auto focus mode
	if (_camera.focusMode == AVCaptureFocusModeLocked)
	{
		if ([_camera lockForConfiguration:nil])
		{
			// restore default focus point
			if ([_camera isFocusPointOfInterestSupported])
			{
				_camera.focusPointOfInterest = CGPointMake(0.5, 0.5);
			}
			
			if ([_camera isFocusModeSupported:AVCaptureFocusModeContinuousAutoFocus])
			{
				[_camera setFocusMode:AVCaptureFocusModeContinuousAutoFocus];
			}
			
			NSLog(@"Focus Mode: Continuos");
		}
	}
}

@end

File: Camera/Swift/DTCameraPreviewController.swift
//
//  DTCameraPreviewController.swift
//  Camera
//
//  Created by Geoff Breemer on 11/10/14.
//  Copyright (c) 2014 Cocoanetics. All rights reserved.
//

import UIKit
import AVFoundation
import Dispatch

@objc(DTCameraPreviewController) class DTCameraPreviewController : UIViewController
{
    @IBOutlet var switchCamButton: UIButton?
    @IBOutlet var toggleTorchButton: UIButton?
    @IBOutlet var snapButton: UIButton?
    
    private var _camera: AVCaptureDevice?
    private var _videoInput: AVCaptureDeviceInput?
    private var _imageOutput: AVCaptureStillImageOutput?
    private var _captureSession: AVCaptureSession?
    private var _videoPreview: DTVideoPreviewView?
    
    deinit
    {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    // MARK: - Internal Methods
    
    func _informUserAboutCamNotAuthorized()
    {
        let alert: UIAlertView = UIAlertView(title: "Cam Access", message: "Access to the camera hardware has been disabled. This disables all related functionality in this app. Please enable it via device settings.", delegate:nil, cancelButtonTitle:"Ok")
        alert.show()
    }
    
    func _informUserAboutNoCam()
    {
        let alert: UIAlertView = UIAlertView(title: "Cam Access", message: "The current device does not have any cameras installed, are you running this in iOS Simulator?", delegate: nil, cancelButtonTitle: "Yes")
        alert.show()
    }
    
    func _setupCamera()
    {
        // get the default camera, usually the one on the back
        _camera = AVCaptureDevice.defaultDeviceWithMediaType(AVMediaTypeVideo)
        
        if (_camera == nil)
        {
            snapButton!.setTitle("No Camera Found", forState: .Normal)
            snapButton!.enabled = false
            
            _informUserAboutNoCam()
            return
        }
        
        _configureCurrentCamera()
        
        // connect camera to input
        var error: NSError?
        _videoInput = AVCaptureDeviceInput(device: _camera, error: &error)
        
        if (error != nil)
        {
            println("Error connecting video input: %@", error!.localizedDescription)
            return
        }
        
        // Create session (use default AVCaptureSessionPresetHigh)
        _captureSession = AVCaptureSession()
        
        if (_captureSession!.canAddInput(_videoInput) == false)
        {
            println("Unable to add video input to capture session")
            return
        }
        
        _captureSession!.addInput(_videoInput)
        
        // add still image output
        _imageOutput = AVCaptureStillImageOutput()
        
        if (_captureSession!.canAddOutput(_imageOutput) == false)
        {
            println("Unable to add still image output to capture session")
            return
        }
        
        _captureSession!.addOutput(_imageOutput)
        
        // set the session to be previewed
        _videoPreview!.previewLayer.session = _captureSession
    }
    
    func _setupCameraAfterCheckingAuthorization()
    {
        // Note: iOS6 check included here before is not required: Swift requires at least iOS7
        
        let authStatus: AVAuthorizationStatus = AVCaptureDevice.authorizationStatusForMediaType(AVMediaTypeVideo)
        
        switch (authStatus)
            {
        case .Authorized:
            _setupCamera()
            
        case .NotDetermined:
            AVCaptureDevice.requestAccessForMediaType(AVMediaTypeVideo, completionHandler: {
                [unowned self]
                (granted: Bool) -> Void in
                
                // background thread, we want to setup on main thread
                dispatch_async(dispatch_get_main_queue(), {
                    if (granted)
                    {
                        self._setupCamera()
                        
                        // need to do more setup because View Controller is already present
                        self._setupCamSwitchButton()
                        self._setupTorchToggleButton()
                        
                        // need to start capture session
                        self._captureSession!.startRunning()
                    }
                    else
                    {
                        self._informUserAboutCamNotAuthorized()
                    }
                })
            })
        case .Restricted, .Denied:
            _informUserAboutCamNotAuthorized()
        }
    }
    
    // applies settings to the currently active camera
    func _configureCurrentCamera()
    {
        // if cam supports AV lock then we want to be able to get out of this
        if (_camera!.isFocusModeSupported(.Locked))
        {
            if (_camera!.lockForConfiguration(nil))
            {
                _camera!.subjectAreaChangeMonitoringEnabled = true
                _camera!.unlockForConfiguration()
            }
        }
    }
    
    // checks if there is an alternative camera to the current one
    func _alternativeCamToCurrent() -> AVCaptureDevice?
    {
        if (_camera == nil)
        {
            // no current camera
            return nil
        }
        
        var allCams: NSArray = AVCaptureDevice.devicesWithMediaType(AVMediaTypeVideo)
        
        for oneCam in allCams
        {
            if (oneCam as? AVCaptureDevice != _camera)
            {
                // found an alternative!
                return oneCam as? AVCaptureDevice
            }
        }
        
        // no alternative cameras present
        return nil
    }
    
    func _captureConnection() -> AVCaptureConnection?
    {
        for connection in _imageOutput!.connections
        {
            for port in (connection as AVCaptureConnection).inputPorts
            {
                if (port.mediaType == AVMediaTypeVideo)
                {
                    return connection as? AVCaptureConnection
                }
            }
        }
        
        // no connection found
        return nil
    }
    
    // update all capture connections for the current interface orientation
    func _updateConnectionsForInterfaceOrientation(interfaceOrientation: UIInterfaceOrientation)
    {
        var captureOrientation: AVCaptureVideoOrientation = DTAVCaptureVideoOrientationForUIInterfaceOrientation(interfaceOrientation)
        
        for connection in _imageOutput!.connections!
        {
            if ((connection as? AVCaptureConnection)!.supportsVideoOrientation)
            {
                (connection as? AVCaptureConnection)!.videoOrientation = captureOrientation
            }
        }
        
        let con: AVCaptureConnection = _videoPreview!.previewLayer.connection
        
        if (con.supportsVideoOrientation)
        {
            con.videoOrientation = captureOrientation
        }
    }
    
    // configures cam switch button
    func _setupCamSwitchButton()
    {
        var alternativeCam: AVCaptureDevice? = _alternativeCamToCurrent()
        
        if (alternativeCam != nil)
        {
            switchCamButton!.hidden = false
            
            var title: NSString
            
            switch (alternativeCam!.position)
                {
            case .Back:
                title = "Back"
            case .Front:
                title = "Front"
            case .Unspecified:
                title = "Other"
            }
            
            switchCamButton!.setTitle(title, forState:.Normal)
        }
        else
        {
            switchCamButton!.hidden = true
        }
    }
    
    // configure torch button for current cam
    func _setupTorchToggleButton()
    {
        if let _camera = _camera {
            if (_camera.hasTorch)
            {
                toggleTorchButton!.hidden = false
            }
        }
        else
        {
            toggleTorchButton!.hidden = true
        }
    }
    
    // MARK: - View Appearance
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        assert(self.view is DTVideoPreviewView, "Wrong root view class \(NSStringFromClass(self.view!.dynamicType)) in \(NSStringFromClass(self.dynamicType))")
        
        _videoPreview = self.view as? DTVideoPreviewView
        
        _setupCameraAfterCheckingAuthorization()
        
        var tap: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: "handleTap:")
        
        self.view.addGestureRecognizer(tap)
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector:"subjectChanged:", name: AVCaptureDeviceSubjectAreaDidChangeNotification, object: nil)
    }
    
    override func viewWillAppear(animated: Bool)
    {
        super.viewWillAppear(animated)
        
        // start session so that we don't see a black rectangle, but video
        _captureSession?.startRunning()

        _setupCamSwitchButton()
        _setupTorchToggleButton()
    }
    
    override func viewDidDisappear(animated: Bool)
    {
        super.viewDidDisappear(animated)
        
        _captureSession!.stopRunning()
    }
    
    override func shouldAutorotate() -> Bool
    {
        if let connection = _videoPreview?.previewLayer.connection
        {
            if connection.supportsVideoOrientation {
                return true
            }
        }
        
        // prevent UI autorotation to avoid confusing the preview layer
        return false
    }
    
    // MARK: - Interface Rotation
    
    // for demonstration all orientations are supported
    override func supportedInterfaceOrientations() -> Int {
        return Int(UIInterfaceOrientationMask.All.rawValue)
    }
    
    override func willRotateToInterfaceOrientation(toInterfaceOrientation: UIInterfaceOrientation, duration: NSTimeInterval)
    {
        
        super.willRotateToInterfaceOrientation(toInterfaceOrientation, duration: duration)
        
        // need to update capture and preview connections
        _updateConnectionsForInterfaceOrientation(toInterfaceOrientation)
    }
    
    // MARK: - Actions
    
    @IBAction func snap(sender: UIButton)
    {
        if (_camera == nil)
        {
            return
        }
        
        // find correct connection
        let videoConnection: AVCaptureConnection? = _captureConnection()
        
        if (videoConnection == nil)
        {
            println("Error: No Video connection found on still image output")
            return
        }
        
        _imageOutput!.captureStillImageAsynchronouslyFromConnection(videoConnection!, completionHandler: { (imageSampleBuffer: CMSampleBuffer!, error: NSError!) -> Void in
            if (error != nil)
            {
                println("Error capturing still image: \(error.localizedDescription)")
                return
            }
            
            let imageData: NSData = AVCaptureStillImageOutput.jpegStillImageNSDataRepresentation(imageSampleBuffer)
            let image: UIImage = UIImage(data: imageData)!
            
            UIImageWriteToSavedPhotosAlbum(image, nil, nil, nil)
            
        })
    }
    
    @IBAction func switchCam(sender: UIButton)
    {
        _captureSession!.beginConfiguration()
        
        _camera = _alternativeCamToCurrent()
        _configureCurrentCamera()
        
        // remove all old inputs
        for input in _captureSession!.inputs
        {
            _captureSession!.removeInput(input as AVCaptureInput)
        }
        
        // add new input
        _videoInput = AVCaptureDeviceInput(device: _camera, error: nil)
        _captureSession!.addInput(_videoInput)
        
        // there are new connections, tell them about current UI orientation
        _updateConnectionsForInterfaceOrientation(self.interfaceOrientation)
        
        _captureSession!.commitConfiguration()
        
        // update the buttons
        _setupCamSwitchButton()
        _setupTorchToggleButton()
    }
    
    @IBAction func toggleTorch(sender: UIButton)
    {
        if (_camera!.hasTorch)
        {
            let torchActive: Bool = _camera!.torchActive
            
            // need to lock, without this there is an exception
            if (_camera!.lockForConfiguration(nil))
            {
                if (torchActive)
                {
                    if (_camera!.isTorchModeSupported(.Off))
                    {
                        _camera!.torchMode = .Off
                    }
                }
                else
                {
                    if (_camera!.isTorchModeSupported(.On))
                    {
                        _camera!.torchMode = .On
                    }
                }
                
                _camera!.unlockForConfiguration()
            }
        }
    }
    
    func handleTap(gesture: UITapGestureRecognizer)
    {
        if (_camera == nil)
        {
            return
        }
        
        if (gesture.state == .Ended)
        {
            // require both focus point and autofocus
            if ( (_camera!.focusPointOfInterestSupported == false) || (_camera!.isFocusModeSupported(.AutoFocus) == false))
            {
                println("Focus Point Not Supported by current camera")
                return
            }
            
            let locationInPreview: CGPoint = gesture.locationInView(_videoPreview)
            let locationInCapture: CGPoint = _videoPreview!.previewLayer.captureDevicePointOfInterestForPoint(locationInPreview)
            
            if (_camera!.lockForConfiguration(nil) == true)
            {
                // this alone does not trigger focussing
                _camera!.focusPointOfInterest = locationInCapture
                
                // this focusses once and then changes to locked
                _camera!.focusMode = .AutoFocus
                
                println("Focus Mode: Locked to Focus Point")
                
                _camera!.unlockForConfiguration()
            }
        }
    }
    
    // MARK: - Notifications
    
    func subjectChanged(notification: NSNotification)
    {
        // switch back to continuous auto focus mode
        if (_camera!.focusMode == .Locked)
        {
            if (_camera!.lockForConfiguration(nil))
            {
                // restore default focus point
                if (_camera!.focusPointOfInterestSupported)
                {
                    _camera!.focusPointOfInterest = CGPointMake(0.5, 0.5)
                }
                
                if (_camera!.isFocusModeSupported(.ContinuousAutoFocus))
                {
                    _camera!.focusMode = .ContinuousAutoFocus
                }
                
                println("Focus Mode: Continuous")
            }
        }
    }
    
}

